File: CPP/core/AIState.cpp
#include "AIState.h"
#include "Enemy.h"
#include <vector>

void AiState::reset_current_behaviour()
{
    m_current_behaviour = nullptr;
}

void AiState::set_parent(Enemy* t_parent)
{
    parent = t_parent;
}

void AiState::add_behaviour(std::unique_ptr<AiBehaviourBase> t_new_behaviour, const int t_weight)
{
    // Set the parent reference
    t_new_behaviour->set_parent(this);

    // Store the module weight
    t_new_behaviour->set_module_weight(t_weight);

    // Add the new behaviour to the list
    m_behaviours.push_back(std::move(t_new_behaviour));
}

void AiState::update()
{
    // Make sure we have a behaviour module selected
    if (m_current_behaviour == nullptr)
    {
        // Get behaviour module random weights
        std::vector<int> weights;
        for (const auto& behaviour : m_behaviours)
        {
            weights.push_back(behaviour->get_module_weight());
        }

        // Select a new behaviour module
        const int selected_index = weighted_random_select(weights);

        // Set current behaviour to the selected behaviour module
        const auto it = std::next(m_behaviours.begin(), selected_index);
        m_current_behaviour = &(*it);

        // Reset the newly selected module
        m_current_behaviour->get()->reset();
    }

    // Execute the current behaviour module
    m_current_behaviour->get()->execute();

    // Check for state change conditions
    for (const auto& condition : m_conditions)
    {
        if (condition->check_triggers())
        {
            parent->set_current_state(condition->get_new_state());
        }
    }
}

File: UNITY/BehaviourModules/Assets/BehaviourModules/Core/AIState.cs
using System;
using System.Linq;
using System.Collections.Generic;
using UnityEngine;
using Random = UnityEngine.Random;

namespace BehaviourModules.Core
{
    /// <summary>
    /// Represents a state for AI agents, managing behavior modules and transitions based on conditions.
    /// </summary>
    [Serializable]
    public class AIState
    {
        [SerializeField] private int stateId;
        [SerializeField] private List<AIBehaviourWeight> behaviours = new List<AIBehaviourWeight>();
        [SerializeField] private List<AICondition> conditions = new List<AICondition>();
        [SerializeField] private AIAgentComponent parent;
        
        private AIBehaviour m_currentBehaviour;
        
        // Selects a random index based on weighted probabilities.
        private int WeightedRandomSelect(List<int> weights)
        {
            int totalWeight = weights.Sum();
            int randomValue = Random.Range(0, totalWeight);
        
            int currentWeight = 0;
            for (int i = 0; i < weights.Count; i++)
            {
                currentWeight += weights[i];
                if (randomValue < currentWeight)
                {
                    return i;
                }
            }
        
            throw new ArgumentException("Failed to select an index.");
        }
        
        // Clears the states current AI behavior, forcing it to select a new one.
        public void ResetCurrentBehaviour()
        {
            m_currentBehaviour = null;
        }
        
        /// Updates the current AI state, selecting and executing behaviors.
        public void UpdateState()
        {
            // Make sure we have a behaviour module selected
            if (!m_currentBehaviour)
            {
                // Select a new behaviour based on weighted probabilities
                List<int> weights = behaviours.Select(behaviour => behaviour.Weight).ToList();
                int newIndex = WeightedRandomSelect(weights);
                m_currentBehaviour = behaviours[newIndex].Behaviour;
                m_currentBehaviour.Reset();
            }
            
            // Execute the current behaviour and check if it has finished
            if (m_currentBehaviour.Execute())
            {
                ResetCurrentBehaviour();
            }
            
            // Check all the condition's triggers to determine if the state should change
            foreach (AICondition condition in conditions.Where(condition => condition.Triggers.Any(trigger => trigger.Check())))
            {
                parent.SetCurrentState(condition.NextState);
            }
        }
    }
}
